pennylane/__init__.py

```diff
@@ -24,6 +24,16 @@
 import pennylane.compiler
 from pennylane.compiler import qjit
 import pennylane.capture
+import pennylane.decomposition
+from pennylane.decomposition import (
+    register_resources,
+    add_decomps,
+    list_decomps,
+    has_decomp,
+    resource_rep,
+    controlled_resource_rep,
+    adjoint_resource_rep,
+)
```

Comment:
Do we want these to all be top-level at this point in time?

Reply:
@isaacdevlugt

Reply:
controlled_resource_rep and adjoint_resource_rep maybe aren't needed top-level, but can't hurt

Reply:
I just worry about promising too much stability, and making the important components harder to locate and access.

Reply:
How about if we only make `register_resources` available at the top level? I think the most common use case would be to define a decomposition rule and using it as an `alt_decomp` or `fixed_decomp`. Things like `has_decomp`, `list_decomps`, and `add_decomps` are most likely going to be used less often. @isaacdevlugt

Reply:
I think my other concern was that if everything is important, nothing is important. We dilute the things users will commonly interact with. Each individual decision to make something top-level seems easy enough, but the bloat creeps up over time.

Reply:
Yeah, the bloat is always something that comes up whenever we're imposing top-level-import on something 😅. How often a user will use these methods definitely comes into play when deciding top-level or not, but the other factor here is how "core" is this feature to PennyLane rather than a specific domain (e.g., kernels, fourier, dla). Transforms are an exception, where they're very interdisciplinary, but at some point the decision was made to keep them separate. 

Here, I see `list_decomps`, `add_decomps`, and `register_resources`, as being interdisciplinary enough to where they should be imported top-level. `has_decomp` is a bit more domain-specific, so that can be tucked inside of `qml.decomposition`.

Let me know what you think!

Reply:
I've condensed the list of things available at top level to: `register_resources`, `list_decomps`, `add_decomps`, and `resource_rep`.